### ISSUE TITLE
ipaclient_setup_nss fails on Ubuntu/Debian

### ISSUE THREAD

**Issue by imp1sh** [freeipa/ansible-freeipa#1137]
I roll out FreeIPA as a client and during that I also use freeipa.ansible_freeipa.ipaclient_setup_nss.

Sadly the process is stuck as the changes seem to result in the OS is in interactive mode and requesting an input:

See my processes, when it's stuck:

```
root       47387   41700  0 13:27 pts/0    00:00:00       /bin/sh -c /usr/bin/python3 /root/.ansible/tmp/ansible-tmp-1693999621.2573123-3214333-199842260532899/AnsiballZ_ipaclient_setup_nss.py && sleep 0
root       47388   47387  0 13:27 pts/0    00:00:01         /usr/bin/python3 /root/.ansible/tmp/ansible-tmp-1693999621.2573123-3214333-199842260532899/AnsiballZ_ipaclient_setup_nss.py
root       48151   47388  0 13:27 pts/0    00:00:00           /usr/bin/perl -w /usr/share/debconf/frontend /sbin/pam-auth-update --package --enable mkhomedir
root       48157   48151  0 13:27 pts/0    00:00:00             /usr/bin/perl -w /sbin/pam-auth-update --package --enable mkhomedir
root       48159   48151  0 13:27 pts/0    00:00:00             whiptail --backtitle Package configuration --title PAM configuration --output-fd 11 --defaultno --yesno -- One or more of the files  /etc/pam.d/common-{auth,account,password,session} have been locally  modified.  Please indicate whether these local changes should be  overridden using the system-provided configuration.  If you decline this option, you will need to manage your system's authentication  configuration by hand.  Override local changes to /etc/pam.d/common-*? 14 77
```

I'm on 1.11.1
The problematic OSes is Ubuntu 20.04 and Debian 12
On CentOS 7 the problem does not occur.

**Comment by imp1sh** [comment]
Just in case anyone else suffers this problem. As a workaround just call `pam-auth-update` on the target system and say `no`.

**Comment by rjeffman** [comment]
What do you mean by "I also use `freeipa.ansible_freeipa.ipaclient_setup_nss.`"?

The roles internal modules are not meant to be used independently of the roles.

**Comment by imp1sh** [comment]
Well that was certainly a bad way to describe it.
To be clear:
I use the role `ipaclient` from the collection. One task of this role is called `Install - Create IPA NSS database` and that makes use of `freeipa.ansible_freeipa.ipaclient_setup_nss`.

**Comment by rjeffman** [comment]
That make things more clear. :-)

I don't remember testing under Debian 12, but Ubuntu 20.04 used to work. 

The failure ocurs when we do a call to a FreeIPA function, so the same issue should also occur, on this node, if the installation was performed through the CLI.

Note that the question is about a file that is locally modified, so I'm not sure if some installed tool is monitoring the file or not.

As far as I can see (without rebuilding my Debian/Ubuntu environments) is that this is a change either in the distro packaging or package tools, or on the node configuration. Either way it does not seem we can do much on ansible-freeipa side.

**Comment by imp1sh** [comment]
Thank you. I will dig around some more and give feedback.

**Comment by jhawkwind** [comment]
Hello, this also happens to Ubuntu 22.04 LTS and Ubuntu 24.04 LTS.
It seems that something is calling `/sbin/pam-auth-update --package --enable mkhomedir` and doesn't "force" so it waits for user interaction..... which there is no interactive session to interact with the whiptail prompt.

Normally if we add `--force`, it would just answer "yes" on that prompt whether to override localized PAM file changes; which would solve the hanging. However, this may make changes to PAM files that may exceed the scope of the IPA client that may lead to other non-standard breakage.

**Comment by jhawkwind** [comment]
Taking a hint from these two articles:
- [Debian Package Configuration System](https://manpages.debian.org/unstable/debconf-doc/debconf.7.en.html)
- [Become User Flags](https://stackoverflow.com/questions/71403187/ansible-become-user-does-not-take-environment-settings)

You can do this in your playbook to have the package manager suppress prompts by adding an environment variable to the sudo process so it passes the ***noninteractive*** front end environment variable that the package manager will use:
```yaml
- name: Apply IPA Client configuration
  hosts: ipaclients
  become: true
  become_flags: "DEBIAN_FRONTEND=noninteractive"
  roles:
    - role: freeipa.ansible_freeipa.ipaclient
      state: present
 ```

This got me over the issue with Ubuntu systems, and probably other Debian boxes. If there are other non-Debian systems, it will likely just ignore that environment variable even if set. So it should be safe to set without any conditions.

The only danger is that if you have any PAM configurations on the files that are custom that you want to keep, they are wiped since the prompt is entirely skipped.

**Comment by rjeffman** [comment]
I'm somewhat drowning in work here, but I'll try to take a look at it in the coming week(s).

I kinda like the environment variable solution.

**Comment by rjeffman** [comment]
So... I give it a try today, and using Ubuntu 24.04 container images I was able to deploy IPA client on it, using the latest release, without any modification.

Can someone provide a configuration that I can use to reproduce this issue?

**Comment by rjeffman** [comment]
Another remark: I had to use ansible-core 2.15, as none of the newer ones were able to run `ansible.legacy.setup` (`gather_facts: true`).

**Comment by jhawkwind** [comment]
> So... I give it a try today, and using Ubuntu 24.04 container images I was able to deploy IPA client on it, using the latest release, without any modification.
> 
> Can someone provide a configuration that I can use to reproduce this issue?

Container images of Ubuntu seems to not have this issue, but a VM or bare-metal install seems to cause this. Not sure what is the difference between the container images and the VM. But a fresh install from ISO or an upgraded one from 20.04, both causes this.

**Comment by rjeffman** [comment]
Man... can't you make my life easier? :rofl:

It'll take some days before I can try it again, using a VM, but since there's different behavior, have you tried to report this to Ubuntu/Debian? (I'll still look for the environment variable workaround, but it would be nice if they are aware of the behavior.)

**Comment by jhawkwind** [comment]
> have you tried to report this to Ubuntu/Debian?
> 
> I'll still look for the environment variable workaround, but it would be nice if they are aware of the behavior.

I believe that is a good idea, maybe they need to build-in the "interactivity detection" rather than rely on the variable or some other method they have now.

**Comment by jhawkwind** [comment]
So, I did some digging around and found that it is the responsibility of the calling method/process of APT to set the environment variable when aptitude is triggered, and it is not responsible for interactivity tests as their "interactivity test" is the environment variable. Which would make sense that a container image would have it hardcoded-in to set that variable globally for you, since containers are generally non-interactive. Even the `ansible.builtin.apt` call sets that environment variable up and is recognized as a bug when it is not set, as shown here: https://github.com/ansible/ansible/issues/21691

So, then the question is whether `ipa-client-install` should do interactivity tests or not? Which leads me to the man page of `ipa-client-install` which has a `--unattended` flag.

Did ansible-freeipa set this flag? If ansible-freeipa set this flag and it gets stuck, then `ipa-client-install` has a bug of not passing "non-interactive state" information to aptitude via the environment variable. However, if ansible-freeipa did not set the `--unattended` flag for `ipa-client-install`, then it would explain the waiting prompt because `ipa-client-install` wouldn't know it is "non-interactive" as that is the state checking method. This is of course presuming `ipa-client-install` is the one calling aptitude to install packages.

**Comment by jhawkwind** [comment]
> @jhawkwind If setting the environment variable fixes the issue, would something like this work?
> 
> ```yaml
> - hosts: all
>   environment:
>     DEBIAN_FRONTEND: "noninteractive"
>   roles:
>   - role: ipaclient
>     state: present
> ```

Yes, it works great; and probably way better with it not escalated.
